File: apps/cli/src/daemon/services/rollback.ts
/**
 * Rollback helpers for worktree creation failure.
 * Implements the rollback sequence from HIGH-LEVEL-DESIGN.md §5.
 */
import { killSession } from "./tmux.js";
import { directPtyRegistry } from "../state/directPtyRegistry.js";
import { worktreeRemove, deleteBranch } from "./git.js";
import { worktreePath as getWorktreePath } from "./paths.js";
import type { ProjectRecord, WorktreeRecord } from "../types.js";

/**
 * Roll back a failed worktree + session creation.
 * Steps (best-effort, each step logged on failure):
 * 1. Kill tmux session (if any)
 * 2. git worktree remove --force
 * 3. git branch -D (if branch was freshly created)
 * 4. Remove worktree record from manifest (caller must persist)
 *
 * @returns Array of error messages for any steps that failed (empty = clean rollback)
 */
export async function rollbackWorktreeCreate(
  project: ProjectRecord,
  worktree: WorktreeRecord,
): Promise<string[]> {
  const errors: string[] = [];

  // 1. Kill sessions (tmux or direct-pty)
  for (const session of worktree.sessions) {
    if (!session.useTmux) {
      // Direct-pty: kill via registry stream (public kill() method).
      directPtyRegistry.get(session.id)?.kill?.();
    } else {
      try {
        await killSession(session.tmuxName);
      } catch (err) {
        errors.push(`tmux kill-session '${session.tmuxName}': ${String(err)}`);
      }
    }
  }

  // 2. Remove git worktree
  const wtPath = getWorktreePath(project.id, worktree.id);
  try {
    await worktreeRemove(project.absolutePath, wtPath);
  } catch (err) {
    errors.push(`git worktree remove '${wtPath}': ${String(err)}`);
  }

  // 3. Delete the branch
  try {
    await deleteBranch(project.absolutePath, worktree.branch);
  } catch (err) {
    errors.push(`git branch -D '${worktree.branch}': ${String(err)}`);
  }

  return errors;
}

File: rust/vst-git/src/rollback.rs
//! Ports `rollback.ts` — rollback helpers for worktree creation failure
//! (HIGH-LEVEL-DESIGN.md §5).

use vst_proc::Tmux;
use vst_types::{ProjectRecord, WorktreeRecord};

use crate::direct_pty::DirectPtyRegistry;
use crate::git::{delete_branch, worktree_remove};
use crate::paths::Paths;

/// Roll back a failed worktree + session creation.
///
/// Steps (best-effort, each step's failure recorded):
/// 1. Kill tmux session (or direct-pty stream) for each session
/// 2. `git worktree remove --force`
/// 3. `git branch -D` (if the branch was freshly created)
///
/// Returns a list of error messages for any steps that failed (empty = clean).
pub async fn rollback_worktree_create(
    project: &ProjectRecord,
    worktree: &WorktreeRecord,
    tmux: &Tmux,
    direct_pty: &DirectPtyRegistry,
    paths: &Paths,
) -> Vec<String> {
    let mut errors: Vec<String> = Vec::new();

    // 1. Kill sessions (tmux or direct-pty)
    for session in &worktree.sessions {
        if !session.use_tmux {
            if let Some(kill) = direct_pty.get(&session.id) {
                kill.kill();
            }
        } else {
            tmux.kill_session(&session.tmux_name);
        }
    }

    // 2. Remove git worktree
    let wt_path = paths.worktree_path(&project.id, &worktree.id);
    if let Err(e) = worktree_remove(&project.absolute_path, &wt_path.to_string_lossy()).await {
        errors.push(format!("git worktree remove '{}': {e}", wt_path.display()));
    }

    // 3. Delete the branch
    if let Err(e) = delete_branch(&project.absolute_path, &worktree.branch).await {
        errors.push(format!("git branch -D '{}': {e}", worktree.branch));
    }

    errors
}
